chain-storage/src/lib.rs

```diff
@@ -607,173 +427,118 @@ where
+        .and_then(|maybe_block| maybe_block.ok_or(Error::BlockNotFound))
+        .map(|block_info_bin| {
+            let mut block_info_reader: &[u8] = &block_info_bin;
+            BlockInfo::deserialize(&mut block_info_reader)
+        })?;
 
-fn for_path_to_nth_ancestor_internal<B, F>(
-    connection: &Connection,
-    index: &ChainStorageIndex<B>,
-    block_hash: &B::Id,
-    distance: u64,
-    mut callback: F,
-) -> Result<BlockInfo<B::Id>, Error>
-where
-    B: Block,
-    F: FnMut(&BlockInfo<B::Id>),
-{
-    let mut cur_block_info =
-        BlockStoreConnection::<B>::get_block_info_internal(connection, index, block_hash)?;
-
-    if distance >= cur_block_info.chain_length {
-        // FIXME: return error
+    if distance >= current.chain_length() {
         panic!(
             "distance {} > chain length {}",
-            distance, cur_block_info.chain_length
+            distance,
+            current.chain_length()
         );
     }
 
-    let target = cur_block_info.chain_length - distance;
-
-    // Travel back through the chain using the back links until we
-    // reach the desired block.
-    while target < cur_block_info.chain_length {
-        // We're not there yet. Use the back link that takes us
-        // furthest back in the chain, without going beyond the
-        // block we're looking for.
-        let best_link = cur_block_info
-            .back_links
-            .iter()
-            .filter(|x| cur_block_info.chain_length - target >= x.distance)
-            .max_by_key(|x| x.distance)
-            .unwrap()
-            .clone();
-        callback(&cur_block_info);
-        cur_block_info = BlockStoreConnection::<B>::get_block_info_internal(
-            connection,
-            index,
-            &best_link.block_hash,
-        )?;
-    }
+    let target = current.chain_length() - distance;
 
-    assert_eq!(target, cur_block_info.chain_length);
-
-    Ok(cur_block_info)
-}
-
-/// Compute the fast link for a block with a given chain_length. Successive
-/// blocks make a chain_length jump equal to differents powers of two, minus
-/// 1, e.g. 1, 3, 7, 15, 31, ...
-fn compute_fast_link(chain_length: u64) -> u64 {
-    let order = chain_length % 32;
-    let distance = if order == 0 { 1 } else { (1 << order) - 1 };
-    if distance < chain_length {
-        chain_length - distance
-    } else {
-        0
+    while target < current.chain_length() {
+        callback(&current);
+        current = info
+            .get(current.parent_id())
+            .map_err(|err| Error::BackendError(Box::new(err)))
+            .and_then(|maybe_block| maybe_block.ok_or(Error::BlockNotFound))
+            .map(|block_info_bin| {
+                let mut block_info_reader: &[u8] = &block_info_bin;
+                BlockInfo::deserialize(&mut block_info_reader)
+            })?;
     }
+
+    Ok(current)
 }
 
 #[cfg(any(test, feature = "with-bench"))]
 pub mod test_utils {
```

Comment:
move this in a different file too so this file becomes smaller and easier to read/review